src/main/java/dev/aziz/grocerystore/repositories/UserPromotionRepository.java

```diff
@@ -0,0 +1,16 @@
+package dev.aziz.grocerystore.repositories;
+
+import dev.aziz.grocerystore.entities.UserPromotion;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.Optional;
+
+@Repository
+public interface UserPromotionRepository extends JpaRepository<UserPromotion, Long> {
+
+    Optional<UserPromotion> findByUserId(Long id);
+
+    List<UserPromotion> findUserPromotionsByUserId(Long id);
```

Comment:
Maybe you can filter the outdated promotions already here with something like AndEndDateLowerThanNow.